In [2]:
import pandas as pd 
import numpy as np

In [3]:
Original_df = pd.read_csv("Airplane Accident.csv")

In [4]:
Original_df.isna().sum()

NtsbNo                    0
EventType                 0
Mkey                      0
EventDate                 0
City                      0
State                  2283
Country                   0
ReportNo               9601
N                         1
HasSafetyRec              0
ReportType             2256
OriginalPublishDate    3443
HighestInjuryLevel     5509
FatalInjuryCount          0
SeriousInjuryCount        0
MinorInjuryCount          0
ProbableCause          3428
EventID                9613
Latitude                  0
Longitude                 0
Make                      0
Model                     0
AirCraftCategory         83
AirportID              4029
AirportName            4271
AmateurBuilt              0
NumberOfEngines        1864
Scheduled              8195
PurposeOfFlight        2351
FAR                      41
AirCraftDamage          738
WeatherCondition       2383
Operator               5806
ReportStatus           2264
RepGenFlag                0
DocketUrl           

In [5]:
Original_df['EventType'].unique()

array(['ACC', 'INC', 'OCC'], dtype=object)

In [6]:
Original_df['WeatherCondition'].unique()

array([nan, 'VMC', 'IMC', 'Unknown'], dtype=object)

In [7]:
Original_df['Model'].value_counts()

Model
172                    387
737                    283
PA28                   165
R44                    163
182                    133
                      ... 
XLT-RG                   1
PHASE II                 1
S-58T                    1
ROTORWAY EXEC 162-F      1
Savage                   1
Name: count, Length: 2569, dtype: int64

In [8]:
Original_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9613 entries, 0 to 9612
Data columns (total 38 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   NtsbNo               9613 non-null   object 
 1   EventType            9613 non-null   object 
 2   Mkey                 9613 non-null   int64  
 3   EventDate            9613 non-null   object 
 4   City                 9613 non-null   object 
 5   State                7330 non-null   object 
 6   Country              9613 non-null   object 
 7   ReportNo             12 non-null     object 
 8   N                    9612 non-null   object 
 9   HasSafetyRec         9613 non-null   bool   
 10  ReportType           7357 non-null   object 
 11  OriginalPublishDate  6170 non-null   object 
 12  HighestInjuryLevel   4104 non-null   object 
 13  FatalInjuryCount     9613 non-null   int64  
 14  SeriousInjuryCount   9613 non-null   int64  
 15  MinorInjuryCount     9613 non-null   i

In [9]:
Original_df['EventType'].unique()

array(['ACC', 'INC', 'OCC'], dtype=object)

In [10]:
num_columns = Original_df.select_dtypes(include=np.number)
cat_columns = Original_df.select_dtypes(include=['object'])
target = Original_df['EventType']

In [11]:
from scipy.stats import chi2_contingency
import pandas as pd

# Initialize lists to store results
p_values = []
interpretasi = []
cols = []

# Iterate through categorical columns
for col in cat_columns.columns:
    # Create a contingency table
    contingency_table = pd.crosstab(cat_columns[col], target)

    # Perform the Chi-Square test of independence
    chi2_stat, p_value, dof, expected = chi2_contingency(contingency_table)

    # Append results
    p_values.append(p_value)
    cols.append(col)

    # Interpret the results
    if p_value < 0.05:
        interpretasi.append('Correlated')
    else:
        interpretasi.append('Not Correlated')

# Create a DataFrame with the results
results_df = pd.DataFrame({
    'column_name': cols,
    'p_value': p_values,
    'result': interpretasi
})

# Display results
results_df


column_name        p_value          result
0                NtsbNo   4.945748e-01  Not Correlated
1             EventType   0.000000e+00      Correlated
2             EventDate   7.231795e-01  Not Correlated
3                  City   1.749900e-53      Correlated
4                 State   1.149119e-03      Correlated
5               Country   0.000000e+00      Correlated
6              ReportNo   3.636432e-01  Not Correlated
7                     N   9.136469e-01  Not Correlated
8            ReportType   2.290478e-03      Correlated
9   OriginalPublishDate   0.000000e+00      Correlated
10   HighestInjuryLevel   9.170080e-04      Correlated
11        ProbableCause   2.574233e-08      Correlated
12                 Make  4.801006e-252      Correlated
13                Model  5.040539e-194      Correlated
14     AirCraftCategory   4.991327e-35      Correlated
15            AirportID   9.782986e-77      Correlated
16          AirportName   3.524004e-05      Correlated
17         AmateurBuilt   2.908848e-39      Correlated
18      NumberOfEngines   0.000000e+00      Correlated
19            Scheduled   4.098115e-95      Correlated
20      PurposeOfFlight   0.000000e+00      Correlated
21                  FAR   0.000000e+00      Correlated
22       AirCraftDamage   0.000000e+00      Correlated
23     WeatherCondition   2.200021e-54      Correlated
24             Operator   9.999895e-01  Not Correlated
25         ReportStatus   3.660634e-11      Correlated
26            DocketUrl   4.931591e-01  Not Correlated
27    DocketPublishDate  6.303304e-256      Correlated

In [22]:
Original_df['AirCraftCategory'].value_counts()

AirCraftCategory
AIR                  8054
HELI                  982
AIR,AIR               132
GLI                   117
BALL                   64
GYRO                   55
WSFT                   44
UNMANNED               24
PPAR                   22
HELI,HELI               6
HELI,AIR                5
ULTR                    5
UNK                     3
,                       3
HELI,UNMANNED           3
AIR,HELI                3
GLI,GLI                 2
AIR,ULTR                1
HELI,UNK                1
AIR,AIR,AIR             1
CSF                     1
UNMANNED,UNMANNED       1
ULTR,AIR                1
Name: count, dtype: int64

In [12]:
import numpy as np
from scipy.stats import f_oneway
import pandas as pd

# Assuming Original_df is already defined

# Select numerical and categorical columns
num_columns = Original_df.select_dtypes(include=np.number)
cat_columns = Original_df.select_dtypes(include=['object'])

# Define target column
target = 'EventType'  # Make sure this is the correct target column

# Initialize lists to store results
p_values = []
interpretasi = []
cols = []

# Iterate through numerical columns
for col in num_columns.columns:
    # Group the numerical column data by the unique values in the target column
    groups = [Original_df[Original_df[target] == category][col] for category in Original_df[target].unique()]
    
    # Perform ANOVA
    f_stat, p_value = f_oneway(*groups)

    # Append results
    p_values.append(p_value)
    cols.append(col)

    # Interpret the results
    if p_value < 0.05:
        interpretasi.append('Correlated')
    else:
        interpretasi.append('Not Correlated')

# Create a DataFrame with the results
results_df = pd.DataFrame({
    'column_name': cols,
    'p_value': p_values,
    'result': interpretasi
})

# Display results
print(results_df)


          column_name       p_value          result
0                Mkey  7.821126e-14      Correlated
1    FatalInjuryCount  1.872789e-04      Correlated
2  SeriousInjuryCount  2.333429e-16      Correlated
3    MinorInjuryCount  6.029840e-03      Correlated
4             EventID           NaN  Not Correlated
5            Latitude  8.744449e-01  Not Correlated
6           Longitude  2.132831e-01  Not Correlated
7         Unnamed: 37           NaN  Not Correlated


City
State
Country
HasSafetyReg
Latitude
Lotitude
Make (manufaktur)
Model
Number of Engines
Weather Condition
FAR (Federal Aviation Regulation)

EventType [Target]

In [13]:
df = Original_df[['City','State','Country','HasSafetyRec','Latitude','Longitude','Make','Model','NumberOfEngines','WeatherCondition','FAR','EventType']]

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9613 entries, 0 to 9612
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   City              9613 non-null   object 
 1   State             7330 non-null   object 
 2   Country           9613 non-null   object 
 3   HasSafetyRec      9613 non-null   bool   
 4   Latitude          9613 non-null   float64
 5   Longitude         9613 non-null   float64
 6   Make              9613 non-null   object 
 7   Model             9613 non-null   object 
 8   NumberOfEngines   7749 non-null   object 
 9   WeatherCondition  7230 non-null   object 
 10  FAR               9572 non-null   object 
 11  EventType         9613 non-null   object 
dtypes: bool(1), float64(2), object(9)
memory usage: 835.6+ KB


In [15]:
from scipy.stats import chi2_contingency
import pandas as pd

# Iterate over categorical features and perform chi-squared test
categorical_features = ['City', 'State', 'Country', 'Make', 'Model', 'WeatherCondition', 'FAR']

p_values = []
results = []

for feature in categorical_features:
    # Create a contingency table (cross-tabulation between target and feature)
    contingency_table = pd.crosstab(Original_df[feature], Original_df['EventType'])
    
    # Perform chi-squared test
    chi2_stat, p_value, dof, expected = chi2_contingency(contingency_table)
    
    p_values.append(p_value)
    
    # Interpret the result based on p-value
    if p_value < 0.05:
        results.append(f'{feature} is correlated with EventType')
    else:
        results.append(f'{feature} is not correlated with EventType')

# Display results
correlation_results = pd.DataFrame({
    'Feature': categorical_features,
    'P-Value': p_values,
    'Interpretation': results
})

correlation_results


Feature        P-Value  \
0              City   1.749900e-53   
1             State   1.149119e-03   
2           Country   0.000000e+00   
3              Make  4.801006e-252   
4             Model  5.040539e-194   
5  WeatherCondition   2.200021e-54   
6               FAR   0.000000e+00   

                                  Interpretation  
0              City is correlated with EventType  
1             State is correlated with EventType  
2           Country is correlated with EventType  
3              Make is correlated with EventType  
4             Model is correlated with EventType  
5  WeatherCondition is correlated with EventType  
6               FAR is correlated with EventType

In [18]:
from scipy.stats import f_oneway

numerical_features = ['Latitude', 'Longitude','NumberOfEngines']  # Add other numerical columns
p_values = []
results = []

for feature in numerical_features:
    # Group the data by the 'EventType' and perform ANOVA
    groups = [Original_df[Original_df['EventType'] == category][feature] for category in Original_df['EventType'].unique()]
    f_stat, p_value = f_oneway(*groups)
    
    p_values.append(p_value)
    
    # Interpret the result based on p-value
    if p_value < 0.05:
        results.append(f'{feature} is correlated with EventType')
    else:
        results.append(f'{feature} is not correlated with EventType')

# Display results
anova_results = pd.DataFrame({
    'Feature': numerical_features,
    'P-Value': p_values,
    'Interpretation': results
})

anova_results


Feature   P-Value                                    Interpretation
0         Latitude  0.874445         Latitude is not correlated with EventType
1        Longitude  0.213283        Longitude is not correlated with EventType
2  NumberOfEngines       NaN  NumberOfEngines is not correlated with EventType

# State Missing Value

In [16]:
df['State'].isna().sum()

2283

In [17]:
df[(df['State'].isna()) & (df['Country'] == 'United States')]

City State        Country  HasSafetyRec   Latitude  \
7510    GULF OF MEXICO   NaN  United States         False  26.014999   
8390    Gulf of Mexico   NaN  United States         False  28.866945   
8409  Charlotte Amalie   NaN  United States         False  18.337499   
8801            Saipan   NaN  United States         False  15.123888   
9268            Saipan   NaN  United States         False  15.118055   

       Longitude    Make    Model NumberOfEngines WeatherCondition  FAR  \
7510  -82.112503  JABIRU  J250-SP               1              VMC  091   
8390  -89.302223   Piper     PA46               1              VMC  091   
8409  -64.967224   Piper     PA23               2              VMC  091   
8801  145.731384   Piper     PA28               1              VMC  091   
9268  145.724166   Piper     PA32               1              VMC  135   

     EventType  
7510       ACC  
8390       ACC  
8409       ACC  
8801       ACC  
9268       ACC

In [18]:
df[(df['State'].notna()) & (df['Country'] != 'United States')]

Empty DataFrame
Columns: [City, State, Country, HasSafetyRec, Latitude, Longitude, Make, Model, NumberOfEngines, WeatherCondition, FAR, EventType]
Index: []

In [19]:
df.loc[:, 'City'] = df['City'].str.lower().str.title().str.replace(r'\bOf\b', 'of', regex=True)

In [20]:
df[(df['State'].isna()) & (df['Country'] == 'United States')]

City State        Country  HasSafetyRec   Latitude  \
7510    Gulf of Mexico   NaN  United States         False  26.014999   
8390    Gulf of Mexico   NaN  United States         False  28.866945   
8409  Charlotte Amalie   NaN  United States         False  18.337499   
8801            Saipan   NaN  United States         False  15.123888   
9268            Saipan   NaN  United States         False  15.118055   

       Longitude    Make    Model NumberOfEngines WeatherCondition  FAR  \
7510  -82.112503  JABIRU  J250-SP               1              VMC  091   
8390  -89.302223   Piper     PA46               1              VMC  091   
8409  -64.967224   Piper     PA23               2              VMC  091   
8801  145.731384   Piper     PA28               1              VMC  091   
9268  145.724166   Piper     PA32               1              VMC  135   

     EventType  
7510       ACC  
8390       ACC  
8409       ACC  
8801       ACC  
9268       ACC

In [21]:
df.loc[(df['City'] == 'Gulf of Mexico'), 'State'] = 'Mexico'
df.loc[(df['City'] == 'Charlotte Amalie'), 'State'] = 'Virgin Islands'
df.loc[(df['City'] == 'Saipan'), 'State'] = 'Northern Mariana Islands'

In [22]:
df[(df['City'] == 'Gulf of Mexico') & (df['Country'] == 'United States')]

City   State        Country  HasSafetyRec   Latitude  \
1249  Gulf of Mexico  Mexico  United States         False  28.348000   
2925  Gulf of Mexico  Mexico  United States         False  28.175447   
3306  Gulf of Mexico  Mexico  United States         False  29.148557   
7510  Gulf of Mexico  Mexico  United States         False  26.014999   
8070  Gulf of Mexico  Mexico  United States         False  28.926708   
8390  Gulf of Mexico  Mexico  United States         False  28.866945   

      Longitude    Make    Model NumberOfEngines WeatherCondition  FAR  \
1249 -86.455400  BOEING  737-7CT               2              VMC  121   
2925 -83.054322  CESSNA     525B               2              VMC  091   
3306 -91.570651    BELL   206-L4               1              VMC  135   
7510 -82.112503  JABIRU  J250-SP               1              VMC  091   
8070 -89.671936    Bell      407               1              VMC  135   
8390 -89.302223   Piper     PA46               1              VMC  091   

     EventType  
1249       ACC  
2925       ACC  
3306       ACC  
7510       ACC  
8070       ACC  
8390       ACC

In [23]:
df[(df['State'].isna()) & (df['Country'] != 'United States')]

City State         Country  HasSafetyRec   Latitude   Longitude  \
3             Oslo   NaN          Norway         False   0.000000    0.000000   
15         Gramado   NaN          Brazil         False -29.367222  -50.860000   
18    Émerainville   NaN          France         False   0.000000    0.000000   
29           Molde   NaN          Norway         False  62.745546    7.267263   
32         Crawley   NaN  United Kingdom         False  51.149218   -0.192406   
...            ...   ...             ...           ...        ...         ...   
9598      Lanseria   NaN    South Africa         False -25.943889   27.921667   
9599      Tooradin   NaN       Australia         False -38.215557  145.423614   
9606       Chennai   NaN           India         False   0.000000    0.000000   
9607         Lagos   NaN         Nigeria         False   0.000000    0.000000   
9612       Ubatuba   NaN          Brazil         False -23.458055  -45.063056   

          Make Model NumberOfEngines WeatherCondition   FAR EventType  
3       BOEING   737             NaN              NaN   NaN       ACC  
15       PIPER  PA42               2              NaN  NUSN       ACC  
18       PIPER  PA28             NaN              NaN  NUSN       ACC  
29      BOEING   737               2              NaN  NUSC       INC  
32      BOEING   737               2              NaN  NUSC       INC  
...        ...   ...             ...              ...   ...       ...  
9598    Cirrus  SR22             NaN              NaN  NUSN       ACC  
9599    Cessna   182               1              NaN  NUSN       ACC  
9606    AIRBUS  A320               2              NaN  NUSC       INC  
9607    BOEING   737             NaN              NaN  NUSC       INC  
9612  Robinson   R44               1              VMC  NUSN       ACC  

[2278 rows x 12 columns]

In [24]:
RowTotal = len(df.index)
USA = df[df['Country'] == 'United States']['Country'].count()
Other = df[df['Country'] != 'United States']['Country'].count()
print(f'From United States country: {USA}')
print(f'From other country: {Other}')
print(f'United State + Other: {USA + Other}')
print(f'All Data: {RowTotal}')

From United States country: 7335
From other country: 2278
United State + Other: 9613
All Data: 9613


In [25]:
df.loc[(df['State'].isna()), 'State'] = 'Undefined'

In [26]:
df['State'].isna().sum()

0

# Number of Engines Missing Values

## BOEING

In [27]:
df.isna().sum()

City                   0
State                  0
Country                0
HasSafetyRec           0
Latitude               0
Longitude              0
Make                   0
Model                  0
NumberOfEngines     1864
WeatherCondition    2383
FAR                   41
EventType              0
dtype: int64

In [28]:
df[(df['NumberOfEngines'].isna()) & (df['Make'] == 'BOEING')][['Make','Model','NumberOfEngines']]

Make      Model NumberOfEngines
3     BOEING        737             NaN
49    BOEING        737             NaN
56    BOEING        767             NaN
67    BOEING  MD 10-30F             NaN
72    BOEING      787-9             NaN
...      ...        ...             ...
9508  BOEING        737             NaN
9521  BOEING        787             NaN
9535  BOEING        737             NaN
9565  BOEING        737             NaN
9607  BOEING        737             NaN

[361 rows x 3 columns]

In [29]:
df[(df['Make'] == 'BOEING') & (df['NumberOfEngines'].isna())]['Model'].unique()

array(['737', '767', 'MD 10-30F', '787-9', '737-700', '777-223',
       '757-231', '787', '727', '747', '737-9', '777', '767-300', 'B777',
       '737-8', '777-3FX (ER)', '747-400', '737 Max 8-200', '737-800',
       '717', '777-319ER', '787-900', 'MD82', '757', '2X6C', '737-8EH',
       '737-4B6', '787-800', 'CV2', '737-8H4'], dtype=object)

Reference:<br>
 - [Boeing website](https://www.boeing.com/commercial/)
 - [Wikipedia](https://en.wikipedia.org/wiki/List_of_Boeing_aircraft)

| Model | Number of Engines | Notes                                      |
|-------|-------------------|--------------------------------------------|
| 717   | 2                 | Twin-engine, regional jet                 |
| 727   | 3                 | Tri-engine, narrow-body jet               |
| 737   | 2                 | Twin-engine, narrow-body jet              |
| 747   | 4                 | Four-engine, wide-body jumbo jet          |
| 757   | 2                 | Twin-engine, narrow-body jet              |
| 767   | 2                 | Twin-engine, wide-body jet                |
| 777   | 2                 | Twin-engine, wide-body jet                |
| 787   | 2                 | Twin-engine, wide-body Dreamliner         |


In [30]:
BoeingEngine = {
    '717': 2,
    '727': 3,
    '737': 2,
    '747': 4,
    '757': 2,
    '767': 2,
    '777': 2,
    '787': 2,
}

In [31]:
for i in df.index:
    if pd.isna(df['NumberOfEngines'][i]) and df['Make'][i] == 'BOEING':
        model = str(df['Model'][i])
        for key in BoeingEngine.keys():
            if key in model:
                df.at[i, 'NumberOfEngines'] = BoeingEngine[key]
                break

In [32]:
df[(df['NumberOfEngines'].isna()) & (df['Make'] == 'BOEING')][['Make','Model','NumberOfEngines']]

Make      Model NumberOfEngines
67    BOEING  MD 10-30F             NaN
4545  BOEING       MD82             NaN
4849  BOEING       2X6C             NaN
8928  BOEING        CV2             NaN

| Make   | Model | Number of Engines | Reference Link                                                                 |
|--------|-------|-------------------|-------------------------------------------------------------------------------|
| BOEING | B777  |2| [Boeing 777](https://en.wikipedia.org/wiki/Boeing_777)                       |
| BOEING | 2X6C  |2| [737 - 2X6C](https://asn.flightsafety.org/wikibase/365686) |
| BOEING | CV2   |12| [Cargo Vessel](https://en.wikipedia.org/wiki/Boeing_Cargo_Air_Vehicle)|
| BOEING | MD 10-30F|3|[MD-10](https://en.wikipedia.org/wiki/McDonnell_Douglas_DC-10#MD-10) |
| BOEING | MD82|2|[MD-82](https://en.wikipedia.org/wiki/McDonnell_Douglas_MD-80#Variants)|



In [33]:
df.loc[(df['Make'] == 'BOEING') & (df['Model'] == 'B777'), 'NumberOfEngines'] = 2
df.loc[(df['Make'] == 'BOEING') & (df['Model'] == '2X6C'), 'NumberOfEngines'] = 2
df.loc[(df['Make'] == 'BOEING') & (df['Model'] == 'CV2'), 'NumberOfEngines'] = 12
df.loc[(df['Make'] == 'BOEING') & (df['Model'] == 'MD 10-30F'), 'NumberOfEngines'] = 3
df.loc[(df['Make'] == 'BOEING') & (df['Model'] == 'MD82'), 'NumberOfEngines'] = 2


In [34]:
df[(df['NumberOfEngines'].isna()) & (df['Make'] == 'BOEING')][['Make','Model','NumberOfEngines']]

Empty DataFrame
Columns: [Make, Model, NumberOfEngines]
Index: []

## Airbus

Reference links:<br>
 - [Airbus website](https://www.airbus.com/en)
 - [Wikipedia](https://en.wikipedia.org/wiki/List_of_Airbus_aircraft)

| Model         | Number of Engines | Notes                                       |
|---------------|-------------------|---------------------------------------------|
| 220          | 2                 | Twin-engine, narrow-body jet                |
| 300          | 2                 | Twin-engine, wide-body jet                  |
| 300 F4-622R  | 2                 | Twin-engine freighter version of A300       |
| 320          | 2                 | Twin-engine, narrow-body jet                |
| 321          | 2                 | Twin-engine, narrow-body jet                |
| 330          | 2                 | Twin-engine, wide-body jet                  |
| 350          | 2                 | Twin-engine, wide-body jet                  |
| 380          | 4                 | Four-engine, double-deck wide-body jet      |
| 350         | 1                 | Single-engine helicopter                    |

In [35]:
AirbusEngine = {
    '220': 2,
    '300': 2,
    '300 F4-622R': 2,
    '320': 2,
    '321': 2,
    '330': 2,
    '350': 2,
    '380': 4,
    '350': 1,
    '332': 2,
    '319': 2
}

In [36]:
for i in df.index:
    if pd.isna(df['NumberOfEngines'][i]) and df['Make'][i] == 'AIRBUS':
        model = df['Model'][i]
        for key in AirbusEngine.keys():
            if key in model:
                df.loc[i, 'NumberOfEngines'] = AirbusEngine[key]
                break

In [37]:
df[(df['NumberOfEngines'].isna()) & (df['Make'] == 'AIRBUS')][['Make','Model','NumberOfEngines']]

Empty DataFrame
Columns: [Make, Model, NumberOfEngines]
Index: []

## CESSNA

In [38]:
df[(df['NumberOfEngines'].isna())][['Make','Model','NumberOfEngines']]

Make     Model NumberOfEngines
0                       CESSNA      172N             NaN
2                       CESSNA      310D             NaN
5     DIAMOND AIRCRAFT IND INC  DA 40 NG             NaN
7                       CESSNA     T210L             NaN
8                        BEECH        58             NaN
...                        ...       ...             ...
9568         MCDONNELL DOUGLAS      MD11             NaN
9569            Hill Group LLC  CCX-2000             NaN
9573                  Champion      7ECA             NaN
9592                     Piper     PA28R             NaN
9598                    Cirrus      SR22             NaN

[1424 rows x 3 columns]

| Model       | Number of Engines | Notes                                           |
|-------------|-------------------|-------------------------------------------------|
| 150 Series  | 1                 | Popular for training and personal aviation      |
| 172 Series  | 1                 | The most produced airplane in history           |
| 182 Series  | 1                 | Known for high performance and range            |
| 185 Series  | 1                 | Utility aircraft, often used for bush flying    |
| 206 Series  | 1                 | Utility aircraft, frequently used in remote areas|
| 210 Series  | 1                 | Pressurized cabin option available (P210 model) |
| 310 Series  | 2                 | First twin-engine model produced by Cessna      |
| 336/337     | 2                 | Push-pull configuration with centerline thrust  |
| 340 Series  | 2                 | Pressurized cabin for business aviation         |
| 401/402     | 2                 | Used for passenger and cargo transport          |
| 414 Series  | 2                 | Pressurized cabin with enhanced comfort         |
| 421 Series  | 2                 | Golden Eagle series, luxury business aircraft   |
| 425 Series  | 2                 | Conquest I, twin turboprop aircraft             |
| 441 Series  | 2                 | Conquest II, twin turboprop for business travel |
| 500/510/525 | 2                 | CitationJet series, business jets               |
| 550/560     | 2                 | Citation series, mid-size business jets         |
| 680 Series  | 2                 | Sovereign, long-range business jet              |
| 750 Series  | 2                 | Citation X, high-speed business jet             |
| 208 Series  | 1                 | Caravan, utility aircraft for cargo/passengers  |
| 425 Series  | 2                 | Conquest, turboprop for light business aviation |


[CESSNA Model Engine Reference](https://en.wikipedia.org/wiki/List_of_Cessna_models)

In [39]:
df[(df['Make'] == 'CESSNA')][['Make','Model','NumberOfEngines']]

Make  Model NumberOfEngines
0     CESSNA   172N             NaN
2     CESSNA   310D             NaN
7     CESSNA  T210L             NaN
12    CESSNA   182B               1
22    CESSNA    172             NaN
...      ...    ...             ...
9460  CESSNA    172               1
9470  CESSNA   U206             NaN
9498  CESSNA  U206G             NaN
9503  CESSNA    150               1
9530  CESSNA    150               1

[1676 rows x 3 columns]

In [40]:
df.loc[:, 'ModelNumber'] = df['Model'].str.extract(r'(\d+)').astype(float)

df.loc[(df['Make'] == 'CESSNA') & (df['ModelNumber'] < 300), 'NumberOfEngines'] = 1
df.loc[(df['Make'] == 'CESSNA') & (df['ModelNumber'] >= 300), 'NumberOfEngines'] = 2

C:\Users\Affan\AppData\Local\Temp\ipykernel_7212\1267209775.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, 'ModelNumber'] = df['Model'].str.extract(r'(\d+)').astype(float)


In [41]:
df = df.drop(['ModelNumber'], axis=1)

In [42]:
df[(df['Make'] == 'CESSNA')][['Make','Model','NumberOfEngines']]

Make  Model NumberOfEngines
0     CESSNA   172N               1
2     CESSNA   310D               2
7     CESSNA  T210L               1
12    CESSNA   182B               1
22    CESSNA    172               1
...      ...    ...             ...
9460  CESSNA    172               1
9470  CESSNA   U206               1
9498  CESSNA  U206G               1
9503  CESSNA    150               1
9530  CESSNA    150               1

[1676 rows x 3 columns]

In [43]:
df.isna().sum()

City                   0
State                  0
Country                0
HasSafetyRec           0
Latitude               0
Longitude              0
Make                   0
Model                  0
NumberOfEngines     1056
WeatherCondition    2383
FAR                   41
EventType              0
dtype: int64

## Missing Values Other

In [44]:
df[(df['NumberOfEngines'].isna())]['Make'].unique()

array(['DIAMOND AIRCRAFT IND INC', 'BEECH', 'BURSEY FREDERICK J', 'Uvify',
       'PIPER', 'REMOS AIRCRAFT GMBH', 'CHRISTEN',
       'ZALL JIHLAVAN AIRPLANES S R O', 'HAWKER', 'ROBINSON',
       'COSTRUZIONI AERONAUTICHE TECNA', 'MD HELICOPTER', 'SCHWEIZER',
       'PILATUS', 'AIR TRACTOR', 'SIKORSKY', 'Textron Beech',
       'BOMBARDIER', 'BELLANCA', 'BELL', 'MCDONNELL DOUGLAS', 'TECNAM',
       'DOUGLAS', 'MARQUART', 'KAWASAKI', 'ATR', 'Cessna', 'LEONARDO',
       'WAG AERO', 'SWEARINGEN', 'CANADAIR', 'CIRRUS',
       'AIRBUS HELICOPTERS', 'BRM AERO SRO', 'ZODIAC',
       'EVEKTOR AEROTECHNIK', 'EMBRAER', 'Just', 'EARTHSTAR',
       'Briten-Norman', 'EXTRA', 'Just Aircraft', 'EUROCOPTER', 'BAE',
       'VIKING AIRCRAFT', 'DE HAVILLAND', 'MURPHY', 'Aergility',
       'De Havilland Canada', 'LEARJET', 'DORNIER', 'BEECRAFT', 'Reims',
       'RANS', 'ROBINSON HELICOPTER', 'AERO COMMANDER', 'ROCKWELL',
       'ZENITH', 'MAULE', 'BLACKSHAPE', 'BRITTEN NORMAN', 'MOONEY',
       'HUGHES', 'T

In [45]:
df[(df['NumberOfEngines'].isna())]['Model'].unique()

array(['DA 40 NG', '58', 'RV-7A', 'IFO', 'PA28', 'REMOS G-3/600',
       'PA-24-250', 'EAGLE II', 'JA-600 SKYLEADER', 'HAWKER 900', 'R44',
       'P2012 TRAVELLER', 'PA-28-140', '369', '269C-1', 'PC12', 'AT301',
       'S76', 'G58', 'PA-28-151', 'Global 7500', 'PA32', '14 19', '206',
       'MD83', 'P2008', 'DC9', 'MA5', 'KH4', 'ATR42', '172', '38', 'PA18',
       'AQ139', 'R66', 'Sport Trainer', 'SA227', 'R-44', 'CL600', 'SR22',
       'AS350', 'BRISTELL S-LSA', '95A55', 'CH601XL', '44', 'EV97',
       'PA46', 'ERJ190', 'CL600 2B19', 'HIGHLANDER', 'E GULL 2000',
       'PA34', 'BN2A-20', '200', 'PA23', 'SuperStol', '269C',
       'JETSTREAM 3100', '35', 'AT502', 'DH 6 - 300', 'DHC6', '55',
       'PA-34-200T', 'REBEL', 'ATLIS 1', 'PA25', '505', 'B200', 'PA28R',
       '9-82', 'DHC-8', '75', '328', '1900', 'PA-25', '300', 'E55',
       'AT401', 'Cessa FR172F', '46', 'C90', 'S12', '24', 'Citabria',
       '22', '407', '500-S', 'Commander', 'CH-801', 'S92', 'S6', 'M7',
       'BK160', 'B

In [46]:
df[(df['NumberOfEngines'].isna())][['Make','Model','NumberOfEngines']]

Make     Model NumberOfEngines
5     DIAMOND AIRCRAFT IND INC  DA 40 NG             NaN
8                        BEECH        58             NaN
13          BURSEY FREDERICK J     RV-7A             NaN
16                       Uvify       IFO             NaN
18                       PIPER      PA28             NaN
...                        ...       ...             ...
9568         MCDONNELL DOUGLAS      MD11             NaN
9569            Hill Group LLC  CCX-2000             NaN
9573                  Champion      7ECA             NaN
9592                     Piper     PA28R             NaN
9598                    Cirrus      SR22             NaN

[1056 rows x 3 columns]

In [47]:
df['ModelNumber'] = df['Model'].str.extract(r'(\d+)').astype(float)
df.loc[df['ModelNumber'] < 300, 'NumberOfEngines'] = 1
df.loc[df['ModelNumber'] >= 300, 'NumberOfEngines'] = 2

In [48]:
df = df.drop(['ModelNumber'], axis=1)

In [49]:
df[(df['NumberOfEngines'].isna())][['Make','Model','NumberOfEngines']]

Make               Model NumberOfEngines
16                Uvify                 IFO             NaN
41             CHRISTEN            EAGLE II             NaN
506            WAG AERO       Sport Trainer             NaN
573        BRM AERO SRO      BRISTELL S-LSA             NaN
665                Just          HIGHLANDER             NaN
...                 ...                 ...             ...
9116      Flight Design                CTSW             NaN
9159       Cub Crafters       CARBON CUB EX             NaN
9247               ACRO  SUPER ACRO SPORT I             NaN
9248  TL Ultralight SRO          STINGSPORT             NaN
9365              HOGAN           Innovator             NaN

[72 rows x 3 columns]

### GULFSTREAM

In [50]:
df[(df['Make']== 'GULFSTREAM')][['Make','Model','NumberOfEngines']]

Make          Model NumberOfEngines
1772  GULFSTREAM           695A               2
2100  GULFSTREAM           690C               2
2463  GULFSTREAM             II             NaN
4084  GULFSTREAM            GIV             NaN
4995  GULFSTREAM            GIV               2
5872  GULFSTREAM  GULFSTREAM200               1
7908  GULFSTREAM            150               1

In [51]:
df.loc[(df['Make'] == 'GULFSTREAM') & (df['Model'] == 'II'), 'NumberOfEngines'] = 2
df.loc[(df['Make'] == 'GULFSTREAM') & (df['Model'] == 'GIV'), 'NumberOfEngines'] = 2

In [52]:
ultralights = ['BRISTELL S-LSA', 'SuperStol', 'Airaile', 'StingSport', 'Ventura']
df.loc[df['Model'].isin(ultralights), 'NumberOfEngines'] = 1

In [53]:
gliders = ['STANDARD CIRRUS', 'ARCUS', 'PHOEBUS']
df.loc[df['Model'].isin(gliders), 'NumberOfEngines'] = 0

In [54]:
helicopters = ['CARBON CUB EX', 'Alta-X', 'Innovator']
df.loc[df['Model'].isin(helicopters), 'NumberOfEngines'] = 1

In [55]:
df[(df['NumberOfEngines'].isna())][['Make','Model','NumberOfEngines']]

Make               Model NumberOfEngines
16                         Uvify                 IFO             NaN
41                      CHRISTEN            EAGLE II             NaN
506                     WAG AERO       Sport Trainer             NaN
665                         Just          HIGHLANDER             NaN
1024                      MURPHY               REBEL             NaN
1309                    BELLANCA            Citabria             NaN
1371                    ROCKWELL           Commander             NaN
2002                     BUCKEYE       DREAM MACHINE             NaN
2077                    CHRISTEN            EAGLE II             NaN
2430                      Cygnet           sea & Sky             NaN
2542                       REIMS                   F             NaN
2545                      Kitfox               Mk IV             NaN
3224            Aero Eli Servizi                YOYO             NaN
3359                    ROTORWAY                 unk             NaN
3648                  Challenger                  II             NaN
3653                     Wagaero           Sportsman             NaN
3846                       Uvify                 IFO             NaN
3950                       drone           Viper Pro             NaN
4151                     CASSUTT                IIIM             NaN
4154               Cassult Racer                IIIM             NaN
4531                     MONNETT             SONERAI             NaN
5038                      Murphy               Rebel             NaN
5070                         BAE                 ATP             NaN
5126           Just Aircraft LLC            Escapade             NaN
5544                  Acroduster                  II             NaN
5563                 SUPERMARINE            SPITFIRE             NaN
5766                      Joplin              Tundra             NaN
6115                  Storm Fury                  RG             NaN
6399        Wag-a-Bond Traveller          Wag-A-Bond             NaN
6505                    ROTORWAY                EXEC             NaN
6631                       Sonex           Tailwheel             NaN
6643                  AFICIONADO       MINI-MAX EROS             NaN
6861            KJOSTAD JORGEN A            WAGABOND             NaN
6879                   Evolution              Revolt             NaN
6965  RAVEN AIRCRAFT CORPPRATION               RAVEN             NaN
7117        Progressive Aerodyne              SEAREY             NaN
7155                      HAWKER            SEA FURY             NaN
7173                 GREG MILLER           GOT ROCKS             NaN
7298           Golden Circle Air              T-Bird             NaN
7464               Flight Design                CTLS             NaN
7498                       SONEX               SONEX             NaN
7562                    Osprey 2           Amphibian             NaN
7584                       Daher              Kodiak             NaN
7595              North American                 SNJ             NaN
7625                        COOK               SONEX             NaN
7933                  Powrachute             AIRWOLF             NaN
8398                        COZY             Mark IV             NaN
8509          SCODA AERONAUTICAL        Super Petrel             NaN
8691                      Culver                 LCA             NaN
8739                        Judd       Challenger II             NaN
8764              WILLIE BENNETT             Glastar             NaN
8775                       STEEN             Skybolt             NaN
8909                 QUICKSILVER               Sport             NaN
8969                       JAMES    Experimental Cub             NaN
8992            S M & T Aircraft                  UH             NaN
9039                      AURORA                 PAV             NaN
9049                DAVID S MEEK           SUN DEVIL             NaN
9116               Flight 

In [56]:
df['NumberOfEngines'].fillna(1, inplace=True)

C:\Users\Affan\AppData\Local\Temp\ipykernel_7212\1750499899.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['NumberOfEngines'].fillna(1, inplace=True)


In [57]:
df[(df['NumberOfEngines'].isna())][['Make','Model','NumberOfEngines']]

Empty DataFrame
Columns: [Make, Model, NumberOfEngines]
Index: []

In [73]:
df['NumberOfEngines'].unique()

array([1, 2, '1', '2', '0', ',', '4', 0], dtype=object)

# FAR Missing Values

In [58]:
df.isna().sum()

City                   0
State                  0
Country                0
HasSafetyRec           0
Latitude               0
Longitude              0
Make                   0
Model                  0
NumberOfEngines        0
WeatherCondition    2383
FAR                   41
EventType              0
dtype: int64

In [59]:
df[(df['Make'] == 'AIRBUS')][['Make','Model','FAR']]

Make     Model   FAR
71    AIRBUS      A330   NaN
80    AIRBUS      A320   NaN
87    AIRBUS       320   129
123   AIRBUS  A320-214   129
145   AIRBUS  A320-214  NUSC
...      ...       ...   ...
9429  AIRBUS      A320  NUSC
9483  AIRBUS      A350   135
9515  AIRBUS      A321  NUSC
9563  AIRBUS      A320  NUSC
9606  AIRBUS      A320  NUSC

[155 rows x 3 columns]

In [60]:
df['FAR'].value_counts()

FAR
091               6372
NUSN              1093
NUSC               619
137                334
135                258
121                232
UNK                197
129                184
PUBU                71
091,091             57
133                 40
NUSC,NUSC           20
121,121             15
137,137              7
NUSN,NUSN            6
135,091              5
121,091              5
129,129              5
NUSC,NUSN            4
107                  4
091,135              4
091,UNK              3
121,129              3
UNK,UNK              3
091K                 3
135,135              2
ARMF                 2
PUBU,PUBU            2
103                  2
091,107              2
129,121              2
,                    2
125                  2
NUSC,UNK             1
129,NUSC             1
NUSC,121             1
091,121              1
121,135              1
091,ARMF             1
NUSC,NUSC,NUSC       1
135,103              1
135,107              1
121,NUSC             1
NUSC,PU

In [61]:
df['FAR'].unique()

array([nan, '091', 'NUSN', 'UNK', '135', 'NUSC', '121', 'PUBU', ',',
       '091,091', '129', '091K', '133', '121,091', 'NUSC,NUSC', '137',
       '091,UNK', '121,121', '121,129', 'NUSC,NUSC,NUSC', 'UNK,UNK',
       '137,137', '125', 'NUSN,NUSN', '129,121', '107', 'NUSC,UNK',
       '129,129', '091,107', 'NUSC,NUSN', '091,135', '103', '135,091',
       '129,NUSC', 'PUBU,PUBU', 'NUSC,121', '091,121', '121,135',
       '091,ARMF', 'ARMF', '135,135', '135,103', '135,107', '121,NUSC',
       'NUSC,PUBU', 'NUSN,129'], dtype=object)

In [62]:
df['FAR'] = df['FAR'].str.split(',').apply(lambda x: ','.join(set(x)) if isinstance(x, list) else x)

In [63]:
df['FAR'].unique()

array([nan, '091', 'NUSN', 'UNK', '135', 'NUSC', '121', 'PUBU', '', '129',
       '091K', '133', '091,121', '137', 'UNK,091', '129,121', '125',
       '107', 'NUSC,UNK', '107,091', 'NUSN,NUSC', '135,091', '103',
       'NUSC,129', 'NUSC,121', '121,091', '135,121', 'ARMF,091', 'ARMF',
       '135,103', '135,107', 'NUSC,PUBU', 'NUSN,129'], dtype=object)

In [64]:
priority_order = ['121', '135', '091', 'NUSC', 'NUSN', 'PUBU', 'UNK']

for i in range(len(df)):
    far_value = df.loc[i, 'FAR']

    if isinstance(far_value, str):
        far_values = far_value.split(',')

        for priority in priority_order:
            if priority in far_values:
                df.loc[i, 'FAR'] = priority
                break

In [65]:
df['FAR'].fillna('091', inplace=True)

C:\Users\Affan\AppData\Local\Temp\ipykernel_7212\363929319.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['FAR'].fillna('091', inplace=True)


In [66]:
df['FAR'].unique()

array(['091', 'NUSN', 'UNK', '135', 'NUSC', '121', 'PUBU', '', '129',
       '091K', '133', '137', '125', '107', '103', 'ARMF'], dtype=object)

# Weather Condition Missing Values

In [67]:
df['WeatherCondition'].unique()

array([nan, 'VMC', 'IMC', 'Unknown'], dtype=object)

In [68]:
df['WeatherCondition'].value_counts()

WeatherCondition
VMC        6846
IMC         292
Unknown      92
Name: count, dtype: int64

In [69]:
df['WeatherCondition'].fillna('Unknown', inplace=True)

C:\Users\Affan\AppData\Local\Temp\ipykernel_7212\4188192386.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['WeatherCondition'].fillna('Unknown', inplace=True)


In [70]:
df['WeatherCondition'].value_counts()

WeatherCondition
VMC        6846
Unknown    2475
IMC         292
Name: count, dtype: int64

In [71]:
df.isna().sum()

City                0
State               0
Country             0
HasSafetyRec        0
Latitude            0
Longitude           0
Make                0
Model               0
NumberOfEngines     0
WeatherCondition    0
FAR                 0
EventType           0
dtype: int64

In [72]:
df.drop_duplicates()

City       State        Country  HasSafetyRec   Latitude  \
0              Anchorage      Alaska  United States         False   0.000000   
1              Galveston       Texas  United States         False  29.304540   
2                Peebles        Ohio  United States         False   0.000000   
3                   Oslo   Undefined         Norway         False   0.000000   
4                Parsons      Kansas  United States         False  37.298300   
...                  ...         ...            ...           ...        ...   
9608            Townsend     Montana  United States         False  46.289165   
9609            Appleton   Minnesota  United States         False  45.227500   
9610          Sacramento  California  United States         False  38.600833   
9611  Santa Maria Island     Florida  United States         False  27.500000   
9612             Ubatuba   Undefined         Brazil         False -23.458055   

       Longitude                  Make   Model NumberOfEngines  \
0       0.000000                CESSNA    172N               1   
1     -94.771416   ROBINSON HELICOPTER     R44               1   
2       0.000000                CESSNA    310D               2   
3       0.000000                BOEING     737               2   
4     -95.263110                  Cozy     Cub               1   
...          ...                   ...     ...             ...   
9608 -111.519447                Cessna     180               1   
9609  -96.004447                Cessna     172               1   
9610 -121.503334  Progressive Aerodyne  SEAREY               1   
9611  -82.746665              Robinson     R44               1   
9612  -45.063056              Robinson     R44               1   

     WeatherCondition   FAR EventType  
0             Unknown   091       ACC  
1                 VMC   091       ACC  
2                 VMC   091       ACC  
3             Unknown   091       ACC  
4             Unknown   091       ACC  
...               ...   ...       ...  
9608              VMC   091       ACC  
9609              VMC   091       ACC  
9610              VMC   091       ACC  
9611              VMC   091       ACC  
9612              VMC  NUSN       ACC  

[9568 rows x 12 columns]